### PR TITLE
roscpp: Adding AsyncSpinner::canStart to check if a spinner can be started

### PR DESCRIPTION
**Pull request by jbohren**
This adds the introspection described in https://github.com/ros/ros_comm/issues/277

Should a PR also be made against hydro?

### CODE REVIEW

File: clients/roscpp/src/libros/spinner.cpp [hunk 1]
```diff
@@ -132,6 +134,18 @@ class AsyncSpinnerImpl
...
 
+bool AsyncSpinnerImpl::canStart()
+{
+  boost::recursive_mutex::scoped_try_lock spinlock(spinmutex);
+
+  if (spinlock.owns_lock()) {
+    return true;
+  }
+
+  return false;
```
Comment (dirk-thomas) [inline clients/roscpp/src/libros/spinner.cpp]: Please simplify the code:

```
return spinlock.owns_lock();
```

File: clients/roscpp/src/libros/spinner.cpp [hunk 2]
```diff
@@ -141,8 +155,15 @@ void AsyncSpinnerImpl::start()
...
   boost::recursive_mutex::scoped_try_lock spinlock(spinmutex);
   if (!spinlock.owns_lock()) {
-    ROS_ERROR("AsyncSpinnerImpl: Attempt to call spin from multiple "
+    ROS_WARN("AsyncSpinnerImpl: Attempt to call spin from multiple "
               "threads.  We already spin multithreaded.");
+    if(callback_queue_ && callback_queue_ != getGlobalCallbackQueue()) {
+      ROS_ERROR("AsyncSpinnerImpl: You're attemping to create"
+                " a second AsyncSpinner with a custom callback"
+                " queue, but another AsyncSpinner already exists,"
+                " so your custom callbac queue will not be serviced"
```
Comment (dirk-thomas) [inline clients/roscpp/src/libros/spinner.cpp]: Spelling: callback

**Comment by jbohren** [comment]
Corrections made.

**Comment by tfoote** [comment]
+1

File: clients/roscpp/src/libros/spinner.cpp [hunk 3]
```diff
@@ -141,8 +150,15 @@ void AsyncSpinnerImpl::start()
...
   if (!spinlock.owns_lock()) {
-    ROS_ERROR("AsyncSpinnerImpl: Attempt to call spin from multiple "
+    ROS_WARN("AsyncSpinnerImpl: Attempt to call spin from multiple "
               "threads.  We already spin multithreaded.");
+    if(callback_queue_ && callback_queue_ != getGlobalCallbackQueue()) {
+      ROS_ERROR("AsyncSpinnerImpl: You're attemping to create"
+                " a second AsyncSpinner with a custom callback"
+                " queue, but another AsyncSpinner already exists,"
+                " so your custom callback queue will not be serviced"
+                " by this AsyncSpinner.");
```
Comment (jbohren) [inline clients/roscpp/src/libros/spinner.cpp]: This error message actually can't be this precise. It's possible that the other `AsyncSpinner` was created with a non-default callback queue, so not even the default queue will be serviced.

**Comment by jbohren** [comment]
I've updated the warning message to better characterize the issue with starting two `AsyncSpinner`s.

**Comment by dirk-thomas** [comment]
+1 Manually merged in e6f48a88202cb5497f27f6dd793553fb1559c232 with a minor whitespace change.

**Comment by dirk-thomas** [comment]
I will cherry-pick the change to the _hydro-devel_ branch once it passed CI - you don't have to create a PR for that.

**Comment by jbohren** [comment]
> I will cherry-pick the change to the hydro-devel branch once it passed CI - you don't have to create a PR for that.

Cool. Thanks, Dirk.

**Comment by dirk-thomas** [comment]
CI passed: http://jenkins.ros.org/view/Idev/job/devel-indigo-ros_comm/24/
Cherry-picked: cb040f2ed6cf26225ba038e881abab400b6c80a2

**Pull request closed without merging**